File: programers/phone_num_prefix/prefix.cpp
#include <string>
#include <vector>
#include <set>
#include <cassert>

using namespace std;

bool solution(vector<string> phone_book) {
    set<string> checkSet{phone_book.begin(), phone_book.end()};
    for (auto phone : phone_book) {
        string prefix;
        for (auto num : phone) {
            prefix += num;

            if (prefix == phone) {
                continue;
            }

            if (checkSet.count(prefix)) {
                return false;
            }
        }
    }
    return true;
}

void testExample() {
    assert(solution({"119", "97674223", "1195524421"}) == false);
    assert(solution({"123","456","789"}) == true);
    assert(solution({"12","123","1235","567","88"}) == false);
}

int main() {
    testExample();

    return 0;
}

File: programers/phone_num_prefix/prefix.py
import pytest

def solution(phone_book):
    last_index = len(phone_book)-1
    phone_book.sort()
    for i in range(last_index):
        if phone_book[i+1].startswith(phone_book[i]):
            return False

    return True

@pytest.mark.parametrize('phone_book, expected', [
    (['119', '97674223', '1195524421'], False),
    (['123','456','789'], True),
    (['12','123','1235','567','88'], False)
])
def test_example(phone_book, expected):
    assert expected == solution(phone_book)
